### ISSUE TITLE
Another question

### ISSUE THREAD

**Issue by VolcanoWCAN** [kehm/ehrblockchain#6]
Hi kehm，
I had another question.
How to understand "Authentication of entities invoking various database operations can be performed in both the blockchain layer and in the database interface"?
My understaing is we use PKI to autheticate in fabric.
We use password to login database.  
I don't understand what's relationship between the blockchain layer and the database interface's function ?
Could you please explain it to me ?

Thank you very much!

**Comment by kehm** [comment]
Hi,
The statement is a reference to Fig 4.1 where we see an illustration of how the blockchain layer can be deployed next to the existing system. The component labelled as "Database interface", which is typically some sort of a DBMS, must be wrapped in software that is embedding one of the Fabric SDKs to incorporate access control mechanisms that are governed by the blockchain.

Local entities (i.e. entities belonging to the same provider) can access the local database through the Database Interface component directly, without being aware of their organization's Blockchain Layer. Meanwhile, external entities must communicate using the Blockchain Layer.
In the former case, the Fabric SDK Database Interface can reject access attempts based on local policies in addition to policies specified in the Blockchain Layer (endorsement policies). 
In the latter case, the Peer nodes in the Blockchain Layer return endorsements directly to the external entity without consulting the local Fabric SDK.

In other words, the thought is that a combination of authentication in the blockchain layer and in the local SDK embedded Database Interface can be configured.

**Comment by VolcanoWCAN** [comment]
Hi!
I appreciate your reply!
I'm a bit confused, I still have the following questions:
1. "In the former case, the Fabric SDK Database Interface can reject access attempts based on local policies in addition to policies specified in the Blockchain Layer (endorsement policies)"
What does endorsement policies mean in this sentence?
Do endorsement policies function as access control policies?
2. Is the blockchain layer or Fabric SDK used for access control?
3. In your system, does the acl mechanism only trigger the action of recording who has access to the blockchain and recording, without implementing the real access control to the database?
4. I have an idea now: I hope that the user will be authenticated in some way before accessing the database. When the user successfully enters the database, the user's operation on the database will also be recorded on the blockchain. Is this idea correct and feasible? Is it similar to the scheme to your idea?
Thank you!

**Comment by VolcanoWCAN** [comment]
I am still a bit confused about the access control in this part.
And I saw in the thesis that the logEntry will be encrypted.
As far as I know, in information in Fabric are Plaintext data without  encryption.  
We can only use channel and private to protect theplain data.
Is my understanding right？

**Comment by kehm** [comment]
Your understanding of data in Fabric is correct: Data is visible for everyone on the channel.
The LogEntry is currently not encrypted. You will find that listed as a "TODO" in the README file:
"TODO: Encrypt LogEntry objects and implement collective authority for key distribution."
The most reasonable way to protect the LogEntry object would probably be to use Fabric Private Data collections.

Regarding your previous questions:
1. An entity is not able to query the state database/blockchain without successful endorsement of the query in the chaincode. The chaincode will only be executed correctly and endorsed if the user is in the ACL. So in that case yes, endorsement policies control access to the blockchain.

2. It is important to note that the proposed system is deployed as an extension to an existing system, with the main purpose of providing standardized auditable logging of events across multiple providers. To achieve this, we simply direct requests from client users to the Fabric SDK embedded application (which is executing in the Database Interface) instead of directly to the DBMS/DB back-end. The SDK application must then decide if a user is authorized by checking the ACL on the blockchain.
We do not, however, propose a way of performing traditional access control for requests bypassing this application, e.g. database queries going straight from a local client to the DBMS. That should be the job of the DBMS/DB back-end solution that is already present in the systems that we target.

3. That is correct. Access control for the underlaying database is implementation specific and should already be in place in the existing system. That also includes protocols for transmitting the actual EHR records to the client.
Note that our system is intended to run next to existing EHR systems in order to alleviate the workload of adopting blockchain solutions, as re-writing existing EHR systems to integrate blockchain technology at a lower level is in most cases not economically feasible.

4. You must invoke chaincode explicitly to check for a matching entry in the ACL and to subsequently log the user action. This can be implemented by having your database interface embedding the Fabric SDK and check for successful chaincode endorsements before returning any data.
